std/random.d

```diff
@@ -1183,86 +1186,230 @@ alias Mt19937_64 = MersenneTwisterEngine!(ulong, 64, 312, 156, 31,
     /// Always `false` (random generators are infinite ranges).
     enum empty = false;
     /// Smallest generated value.
-    enum UIntType min = seeds_.length == 1 ? 1 : 0;
+    enum UIntType min = _state.length == 1 ? 1 : 0;
     /// Largest generated value.
     enum UIntType max = UIntType.max;
 
 
   private:
-    enum size = bits / 32;
-
-    static if (bits == 32)
-        UIntType[size] seeds_ = [2_463_534_242];
-    else static if (bits == 64)
-        UIntType[size] seeds_ = [123_456_789, 362_436_069];
-    else static if (bits == 96)
-        UIntType[size] seeds_ = [123_456_789, 362_436_069, 521_288_629];
-    else static if (bits == 128)
-        UIntType[size] seeds_ = [123_456_789, 362_436_069, 521_288_629, 88_675_123];
-    else static if (bits == 160)
-        UIntType[size] seeds_ = [123_456_789, 362_436_069, 521_288_629, 88_675_123, 5_783_321];
-    else static if (bits == 192)
-    {
-        UIntType[size] seeds_ = [123_456_789, 362_436_069, 521_288_629, 88_675_123, 5_783_321, 6_615_241];
+    // Legacy 192 bit uint hybrid counter/xorshift.
+    enum bool isLegacy192Bit = UIntType.sizeof == uint.sizeof && nbits == 192;
+
+    // Shift magnitudes.
+    enum a = (sa < 0 ? -sa : sa);
+    enum b = (sb < 0 ? -sb : sb);
+    enum c = (sc < 0 ? -sc : sc);
+
+    // Shift expressions to mix in.
+    enum shiftA(string expr) = `((`~expr~`) `~(sa > 0 ? `<< a)` : ` >>> a)`);
+    enum shiftB(string expr) = `((`~expr~`) `~(sb > 0 ? `<< b)` : ` >>> b)`);
+    enum shiftC(string expr) = `((`~expr~`) `~(sc > 0 ? `<< c)` : ` >>> c)`);
+
+    enum N = nbits / (UIntType.sizeof * 8);
+
+    // For N <= 2 it is strictly worse to use an index.
+    // Informal third-party benchmarks suggest that for `ulong` it is
+    // faster to use an index when N == 4. For `uint` we err on the side
+    // of not increasing the struct's size and only switch to the other
+    // implementation when N > 4.
+    enum useIndex = !isLegacy192Bit && (UIntType.sizeof >= ulong.sizeof ? N > 3 : N > 4);
+    static if (useIndex)
+    {
+        enum initialIndex = N - 1;
+        uint _index = initialIndex;
+    }
+
+    static if (N == 1 && UIntType.sizeof <= uint.sizeof)
+    {
+        UIntType[N] _state = [cast(UIntType) 2_463_534_242];
+    }
+    else static if (isLegacy192Bit)
+    {
+        UIntType[N] _state = [123_456_789, 362_436_069, 521_288_629, 88_675_123, 5_783_321, 6_615_241];
         UIntType       value_;
     }
+    else static if (N <= 5 && UIntType.sizeof <= uint.sizeof)
+    {
+        UIntType[N] _state = [
+            cast(UIntType) 123_456_789,
+            cast(UIntType) 362_436_069,
+            cast(UIntType) 521_288_629,
+            cast(UIntType)  88_675_123,
+            cast(UIntType)   5_783_321][0 .. N];
+    }
     else
     {
-        static assert(false, "Phobos Error: Xorshift has no instantiation rule for "
-                             ~ to!string(bits) ~ " bits.");
+        UIntType[N] _state = ()
```

Comment:
This whole delegate is not covered. Maybe some tests for UIntType as `ulong` must be added.

Reply:
This PR tests Xorshift64_64 and Xorshift128_64 with UIntType = ulong. Could the coverage report be confused because this delegate is only run at compile time?

Reply:
Sorry i missed the version(unittest) declarations and how they're used. I don't know if the comment about cov decreasing was relevant then.